aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,126 @@
+//
+// There is something of a pattern to writing these; check THAllocator.{h,cpp}
+// for some examples.
+
+using DeleterFnPtr = void(*)(void*);
+using SupervisorPtr = std::unique_ptr<void, DeleterFnPtr>;
+
+// Does not delete anything
+AT_API void deleteNothing(void*);
+// Mints a SupervisorPtr that doesn't do anything.  You must
+// use this, not a nullptr, when you want a view.
+AT_API SupervisorPtr nonOwningSupervisorPtr();
+
+struct SupervisedPtr {
+  // Lifetime tied to supervisor_
+  void* data_;
+  SupervisorPtr supervisor_;
+  Device device_;
+  SupervisedPtr() : data_(nullptr), supervisor_(nonOwningSupervisorPtr()), device_(kCPU) {}
+  SupervisedPtr(void* data, SupervisorPtr&& supervisor, Device device)
+    : data_(data), supervisor_(std::move(supervisor)), device_(device) {}
+  void* operator->() const { return data_; }
+  void* get() const { return data_; }
+  operator bool() { return data_ || supervisor_; }
+};
+
+// NB: Device is NOT tested for here; a CUDA nullptr is as much a nullptr as a
+// CPU nullptr
+
+inline bool operator==(const at::SupervisedPtr& sp, std::nullptr_t) noexcept {
+  return sp.data_ == nullptr && sp.supervisor_ == nullptr;
+}
+inline bool operator==(std::nullptr_t, const at::SupervisedPtr& sp) noexcept {
+  return sp.data_ == nullptr && sp.supervisor_ == nullptr;
+}
+inline bool operator!=(const at::SupervisedPtr& sp, std::nullptr_t) noexcept {
+  return sp.data_ != nullptr || sp.supervisor_ != nullptr;
+}
+inline bool operator!=(std::nullptr_t, const at::SupervisedPtr& sp) noexcept {
+  return sp.data_ != nullptr || sp.supervisor_ != nullptr;
+}
+
+// Note [raw_allocate/raw_deallocate and Thrust]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Thrust's support for custom allocators requires us to write something
+// like this:
+//
+//  class ThrustAllocator {
+//    char* allocate(size_t);
+//    void deallocate(char*, size_t);
+//  };
+//
+// This is not good for our unique_ptr based allocator interface, as
+// there is no way to get the deleter from our allocator to the
+// deletion site in Thrust.  Indeed, if every pointer is getting a
+// *fresh* deleter, we truly would have no choice except to maintain
+// a map from pointers to deleters.  This is bad.
+//
+// So, we observe that not *all* deleters actually have lots of
+// different deleters; some of them actually always return the same
+// deleter every time.  In this case, we can support the "raw"
+// allocate and deallocate interface.  This is what
+// maybeGlobalBoundDeleter signifies.  By default, it returns the
+// default (empty) BoundDeleter, which means that the raw interface
+// is not implemented.  Be sure to implement it whenever possible.
+
 struct Allocator {
   virtual ~Allocator() {}
-  virtual void* allocate(size_t n) const = 0;
-  virtual void deallocate(void* ptr) const = 0;
+  virtual at::SupervisedPtr allocate(size_t n) const = 0;
+
+  // If this returns a non nullptr, it means that allocate()
```

Comment:
Why overload the deleter to indicate if the interface is implemented?  Just have a "has_raw_interface" or something?

Reply:
Normally, the deleter is supposed to come in the `SupervisedPtr` package (attached to the supervisor), but for the raw interface you don't have a SupervisedPtr, so you need to get it from somewhere. So `has_raw_interface` (which ostensibly returns a bool) doesn't give you enough information.

But perhaps your suggestion is that, the function pointer deleter attached to the supervisor is always the same, no matter what the supervisor itself is. This is certainly true in all of the code here, and I can't really think of a crazy situation where an allocator may return one or another function pointer. But I would claim that if an `Allocator` gives you a way to get at the deleter, it would only be used in the raw-API case.

I suppose you COULD do this:

```
struct Allocator {
  virtual ~Allocator() {}
  virtual std::pair<void*, void*> vallocate(size_t n);
  virtual DeleterFnPtr get_deleter();
  at::SupervisedPtr allocate(size_t n) {
    std::tie(data, supervisor) = vallocate(n);
    return {data, {supervisor, get_deleter()}};
  }
}
```

But this seems a bit more oblique for no good reason.